### PR TITLE
Hunting down a slow Rails request

### PR DESCRIPTION
**Pull request by mauro-oto**
Hey @etagwerker, @schmierkov, 

I wrote a blog post on how I fixed the problem with a slow Rails API request.

I am not sure whether we should use the `EmailAccount` and `EmailAccountSync` model names, or we should change that to be something else, just for privacy/secrecy/legal matters. Maybe we could ask Preston for permission first. What do you think? 

Please check it out, thanks!

### CODE REVIEW

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 1]
```diff
@@ -0,0 +1,97 @@
+---
+layout: post
+title: "Hunting Down a Slow Rails Request"
+date: 2016-03-04 11:03:00
+categories: ["rails", "performance"]
+author: "mauro-oto"
+---
+
+We have recently began using [Skylight](https://www.skylight.io) in production
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: @mauro-oto I think it'd sound more natural if you said: "Recently we started using... "

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+for one of our client's Rails applications, as an attempt to try to improve the
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: "for one of our clients." sounds a bit better. 
I think that "in an attempt to" sounds bettern than "as an attempt"

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+performance of some of the more critical API endpoints.
+
+Skylight reports on both time taken, with a breakdown of time taken per SQL
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: @mauro-oto This sounds a bit weird, could you refactor it to something like this? 

Skylight reports: 

* The time each request takes to complete
* A breakdown of the time it takes per SQL query
* Object allocations per request

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+query, and also on object allocations per request. I noticed an unusually large
+amount of allocated objects for one request:
+
+![Skylight report](/blog/assets/images/high-object-allocation.png)
+
+This request would take anywhere from 400ms to 3000ms to respond, which is WAY,
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: @mauro-oto I'd just say "*WAY* too long"

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+To get the amount of allocated strings, I used the following snippet:
+
+```ruby
+strings = {}
+ObjectSpace.each_object(String) { |str| strings[str.class] += 1 }
+pp strings
+```
+
+As you can see, there are 700k+ allocated strings, and this increase in object
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: @mauro-oto I'd separate this into "As you can see, there are 700k+ allocated strings. It's very likely that this amount of allocations is the culprit for the slow performance."

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+allocation is the likely culprit for the slow request.
+
+I dug a bit into the `EmailAccount` model and its associations, and noticed that
+this model has a one-to-many association with the `EmailAccountSync` model.
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: * "this model had a one-to-many association"

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+However, we were only returning one of model `EmailAccountSync`, supposedly, by
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: @mauro-oto I'd say "Supposedly we were only returning one `EmailAccountSync` record by using: "

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+As it turns out, the `has_one :latest_sync` relationship was not being
+respected by the query performed in the `index` action, and thus:
+
+```ruby
+EmailAccountSync Load (21.9ms)  SELECT "email_account_syncs".* FROM "email_account_syncs"  WHERE "email_account_syncs"."email_account_id" IN (24, 23, 22, 21, 20, 19, 18, 17, 16, 15)  ORDER BY email_account_syncs.last_attempt_at desc
+```
+
+All `EmailAccountSync` instances for the loaded `EmailAccount` instances get
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: I'd change this to: 

"This meant that all `EmailAccountSync` instances were getting loaded for each `EmailAccount` record. The `limit` that should've been in the SQL statement was not there."

**Comment by etagwerker** [comment]
@mauro-oto Just reviewed this PR. Great work! :beers:

**Comment by mauro-oto** [comment]
@etagwerker just addressed your comments. 

I also made `ul`s the same font size as `p`s for the post excerpts, as the lists looked bigger than paragraphs in the blog index page, see the first post in http://www.ombulabs.com/blog. 

Please check it out, thanks!

File: _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  Parameters: {"results_per_page"=>"30"}
+
+[[String, 150295], [ActiveSupport::JSON::Encoding::JSONGemEncoder::EscapedString, 2110], [Arel::Nodes::SqlLiteral, 20], [Arel::Nodes::BindParam, 10], [ActiveSupport::StringInquirer, 1]]
+### JSON rendering happens ###
+[[String, 157847], [ActiveSupport::JSON::Encoding::JSONGemEncoder::EscapedString, 4220], [Arel::Nodes::SqlLiteral, 30], [Arel::Nodes::BindParam, 12], [ActiveSupport::StringInquirer, 1]]
+
+Completed 200 OK in 178ms (Views: 5.6ms | ActiveRecord: 2.3ms)
+```
+
+The object allocation is down by a huge amount, and the request now takes around
```
Comment (etagwerker) [inline _posts/2016-03-04-hunting-down-a-slow-rails-request.markdown]: "the request now takes from 50s to 200ms to response. This is not yet completely efficient, but it is a much more acceptable performance."

**Comment by mauro-oto** [comment]
@etagwerker thanks! Let me know if you have any other comments. :tada:

**Comment by etagwerker** [comment]
@mauro-oto :+1: :tada:

**Pull request merged**